src/messaging/ExchangeMgr.h

```diff
@@ -0,0 +1,266 @@
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE If the state is not equal to
+     *          kState_NotInitialized.
+     *  @retval #CHIP_NO_ERROR On success.
+     *
+     */
+    CHIP_ERROR Init(SecureSessionMgrBase * sessionMgr);
+
+    /**
+     *  Shutdown the ExchangeManager. This terminates this instance
+     *  of the object and releases all held resources.
+     *
+     *  @note
+     *     The application should only call this function after ensuring that
+     *     there are no active ExchangeContext objects. Furthermore, it is the
+     *     onus of the application to de-allocate the ExchangeManager
+     *     object after calling ExchangeManager::Shutdown().
+     *
+     *  @return #CHIP_NO_ERROR unconditionally.
+     */
+    CHIP_ERROR Shutdown();
+
+    /**
+     *  Creates a new ExchangeContext with a given peer CHIP node specified by the peer node identifier.
+     *
+     *  @param[in]    peerNodeId    The node identifier of the peer with which the ExchangeContext is being set up.
+     *
+     *  @param[in]    appState      A pointer to a higher layer object that holds context state.
+     *
+     *  @return   A pointer to the created ExchangeContext object On success. Otherwise NULL if no object
+     *            can be allocated or is available.
+     */
+    ExchangeContext * NewContext(const uint64_t & peerNodeId, void * appState = nullptr);
+
+    /**
+     *  Find the ExchangeContext from a pool matching a given set of parameters.
+     *
+     *  @param[in]    peerNodeId    The node identifier of the peer with which the ExchangeContext has been set up.
+     *
+     *  @param[in]    appState      A pointer to a higher layer object that holds context state.
+     *
+     *  @param[in]    isInitiator   Boolean indicator of whether the local node is the initiator of the exchange.
+     *
+     *  @return   A pointer to the ExchangeContext object matching the provided parameters On success, NULL on no match.
+     */
+    ExchangeContext * FindContext(uint64_t peerNodeId, void * appState, bool isInitiator);
+
+    /**
+     *  Register an unsolicited message handler for a given profile identifier. This handler would be
+     *  invoked for all messages of the given profile.
+     *
+     *  @param[in]    protocolId     The profile identifier of the received message.
+     *
+     *  @param[in]    handler       The unsolicited message handler.
+     *
+     *  @param[in]    appState      A pointer to a higher layer object that holds context state.
+     *
+     *  @retval #CHIP_ERROR_TOO_MANY_UNSOLICITED_MESSAGE_HANDLERS If the unsolicited message handler pool
+     *                                                             is full and a new one cannot be allocated.
+     *  @retval #CHIP_NO_ERROR On success.
+     */
+    CHIP_ERROR RegisterUnsolicitedMessageHandler(uint32_t protocolId, ExchangeContext::MessageReceiveFunct handler,
```

Comment:
Should we use delegate pattern for defining/adding these callbacks?

Reply:
No, delegate pattern only works when the callback functions are well-known. The unsolicited message handler various among different protocols, even within the same protocols.